ml-agents/mlagents/trainers/settings.py

```diff
@@ -312,6 +294,168 @@ def apply(self, key: str, env_channel: EnvironmentParametersChannel) -> None:
+
+    @staticmethod
+    def _check_lesson_chain(lessons, parameter_name):
+        num_lessons = len(lessons)
+        for index, lesson in enumerate(lessons):
+            if index < num_lessons - 1 and lesson.completion_criteria is None:
+                raise TrainerConfigError(
+                    f"A non-terminal lesson does not have a completion_criteria for {parameter_name}."
+                )
+
+    @staticmethod
+    def _sampler_from_config(
+        environment_parameter_config: Mapping
+    ) -> Optional[ParameterRandomizationSettings]:
+        """
+        Returns a ParameterRandomizationSettings when the environment_parameter_config
+        argument corresponds to a sampler and None otherwise.
+        """
+        if environment_parameter_config is None:
+            return None
+        if isinstance(environment_parameter_config, (float, int)):
+            sampler = ConstantSettings(value=float(environment_parameter_config))
+            return sampler
+        elif "sampler_type" in environment_parameter_config:
+            # This is the non-constant sampler case
+            enum_key = ParameterRandomizationType(
+                environment_parameter_config["sampler_type"]
+            )
+            t = enum_key.to_settings()
+            sampler = strict_to_cls(
+                environment_parameter_config["sampler_parameters"], t
+            )
+            return sampler
+        return None
+
+    @staticmethod
+    def structure(d: Mapping, t: type) -> Dict[str, "EnvironmentParameterSettings"]:
+        """
+        Helper method to structure a Dict of EnvironmentParameterSettings class. Meant
+        to be registered with cattr.register_structure_hook() and called with
+        cattr.structure().
+        """
+        if not isinstance(d, Mapping):
+            raise TrainerConfigError(
+                f"Unsupported parameter environment parameter settings {d}."
+            )
+        d_final: Dict[str, EnvironmentParameterSettings] = {}
+        for environment_parameter, environment_parameter_config in d.items():
+            maybe_sampler = EnvironmentParameterSettings._sampler_from_config(
+                environment_parameter_config
+            )
+            if maybe_sampler is not None:
+                d_final[environment_parameter] = EnvironmentParameterSettings(
+                    lessons=[
+                        Lesson(
+                            completion_criteria=None,
+                            sampler=maybe_sampler,
+                            name=environment_parameter,
+                        )
+                    ]
+                )
+            elif "curriculum" in environment_parameter_config:
+                # This is the curriculum case
+                lessons: List[Lesson] = []
+                for lesson_dict in environment_parameter_config["curriculum"]:
+                    # a lesson_dict contains a single lesson with the name of the lesson as key
```

Comment:
I think you can make a `Lesson.structure` method with the contents of this method, then call `cattr.register_structure_hook(Lesson, <this_method>)`. Then you don't need to iterate through the lessons, `cattr` will automatically handle the List and try to structure each element one by one, since it knows that `EnvironmentParameterSettings` contains a `List[Lesson]`.

Reply:
done